modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -86,6 +86,14 @@ protected BulkRequest buildBulk(Iterable<SearchHit> docs) {
         Map<String, Object> scriptCtx = null;
 
         for (SearchHit doc : docs) {
+            if (doc.isSourceEmpty()) {
+                /*
+                 * Either the document didn't store _source or we didn't fetch it for some reason. Since we don't allow the user to
+                 * change the "fields" part of the search request it is unlikely that we got here because we didn't fetch _source.
+                 * Thus the error message assumes that it wasn't stored.
+                 */
+                throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
```

Comment:
Just tested and the following creates a document in Elasticsearch:

```
PUT test4/doc/1
{
  
}
```

Response:
```
{
  "_index": "test4",
  "_type": "doc",
  "_id": "1",
  "_version": 1,
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "created": true
}
```

So you can definitely create a document with no fields in ES. In this case I think it's wrong for us to throw any exception if we read a document from and index which has no source when re-indexing, we should in fact index it into the new index like any other document. So although I like the checks you added to stop source filtering when re-indexing, I don't think we should be throwing an IllegalArgumentException here.